### ISSUE TITLE
Error in LESS code causes silent failure?

### ISSUE THREAD

**Issue by joshribakoff** [plus3network/gulp-less#171]
Normally in the past the only issues I've had with this package is that it crashes the gulp watch on error. I have not changed anything recently, but I just noticed it fails silently instead of crashing now, when I give it invalid LESS inputs:

I start with an empty folder
```
 ll dist
total 0

```

I have a gulpfile which is roughly this:

```
gulp.task('less', function () {
    return gulp.src('./test.less')
        .pipe(less({
            
        }))
        .pipe(gulp.dest('./dist'));
});
```

contents of test.less:
```
body {
  rgba(179, 229, 252, 0.50);
}
```

back to console, I try to compile:
```
gulp less
[22:06:49] Warning: gulp version mismatch:
[22:06:49] Global gulp is 3.9.0
[22:06:49] Local gulp is 3.8.11
[22:06:50] Using gulpfile ~/www/isite/snmenus2/gulpfile.js
[22:06:50] Starting 'less'...

```
That's the full output. I do not get any further error messages (am not checking return codes). No other output occurs. I will list my `dist` dir again after running `gulp less`:

```
ll dist
total 0

```

If I fix the invalid syntax by prefixing it with "background-color:", it works as expected. I get an output of a "test.css" in my dir, and this gets printed to console:

```
[22:08:01] Finished 'less' after 33 ms

```

Those are my actual results. My expected results would to see a LESS error message specifying the line number that is incorrect, printed to stderr of my console.

Thanks!

**Comment by psolbach** [comment]
+1, just noticed it.

**Comment by stevelacy** [comment]
Related to #171
Less is consuming all the errors.

**Comment by joshribakoff** [comment]
@stevelacy  less is not consuming the errors, gulp-less is. Also related to #118 - they recommend checking output of lessc, as you can see, lessc correctly reports the errors:

![selection_433](https://cloud.githubusercontent.com/assets/4021306/9183733/ffd89100-3f65-11e5-9205-ca8b0a0d20b1.png)

**Comment by stevelacy** [comment]
@joshribakoff gulp-less is a through stream. Any functions inside it modify the file, throw errors, etc. In this case less is actually modifying the error with it's JS API inside it's own promise scope.

I manually took out the less render function and replaced it with a normal file parse. It worked correctly without consuming the errors.

But anyway, not to argue on the semantics of the issue I will look into it when I have free time.

**Comment by mattsawyer77** [comment]
I'm seeing this problem in gulp-less 3.3.0 (I just checked to make sure https://github.com/plus3network/gulp-less/commit/be18b01e7087db7336a35bf2924db3a0e84984fa is in my copy of index.js). 

Has there been a regression, or is there some other way to capture a parsing error?

**Comment by contra** [comment]
@mattsawyer77 Code?

**Comment by mattsawyer77** [comment]
@contra My less task is essentially the following:
```javascript
gulp.task('less', function() {
    var dest = uiRoot + '/css';
    gulpUtil.log('compiling LESS...');

    return gulp.src(sourcePaths.less)
    .pipe(less({
        sourceMap: true
    }))
    .pipe(gulp.dest(dest));
});
```
I can easily repro this by inserting some invalid characters into a .less file, which results in
```
Potentially unhandled rejection [2] Unrecognised input. Possibly missing opening '{' in file ***filename here*** line
no. 1
```
but gulp abruptly stops executing at that point, and I get a 0 exit code. I tried instrumenting the stream with .on('error') and returning a promise which rejects if the stream errors, but the error handler was never called.

If I remove the invalid characters, gulp works fine, and I can see that my less task finishes.

**Comment by mattsawyer77** [comment]
should I file a new issue for this? Or do you think I'm potentially doing something wrong here?

**Comment by contra** [comment]
```js
gulp.task('less', function() {
    return gulp.src(sourcePaths.less)
      .pipe(less({
          sourceMap: true
      }).on('error', console.log))
      .pipe(gulp.dest(uiRoot + '/css'));
});
```

Does this log the error for you?

**Comment by mattsawyer77** [comment]
@contra it does, thank you!

**Comment by lwr** [comment]
+1 with @mattsawyer77 

problem still exists in 3.3.2

I am confused if this problem should be fix in future release?

The workaround https://github.com/stevelacy/gulp-less/issues/171#issuecomment-284468687 works
```js
gulp.task('less', cb => {
     gulp.src(lessSourcePaths)
         .pipe(less(...))
         .on('error', cb)
         .pipe(gulp.dest(cssDestRoot))
         .on('end', cb);
}
```
But it still a very strange behavior that less consumes all errors and we even got a **success** exit code for gulp process

**Comment by lwr** [comment]
we are now gulp 4.0.0 + gulp-less 4.0.0

problem still exists

does not work: error printed out but task succeed
```js
gulp.task('less'), () => {
    return gulp.src(lessSourcePaths)
        .pipe(less(...))
        .pipe(gulp.dest(cssDestRoot))
}
```

not works neither: error totally lost
```js
gulp.task('less'), cb => {
    return gulp.src(lessSourcePaths)
        .pipe(less(...))
        .on('error', cb)
        .pipe(gulp.dest(cssDestRoot))
}
```

not works neither: error totally lost
```js
gulp.task('less'), cb => {
    gulp.src(lessSourcePaths)
        .pipe(less(...))
        .on('error', cb)
        .pipe(gulp.dest(cssDestRoot))
        .on('end', cb);
}
```

finally this workaround works
```js
gulp.task('less'), cb => {
    let succeed;
    gulp.src(lessSourcePaths)
        .on('end', () => { succeed = true })
        .pipe(less(...))
        .on('end', () => { succeed && cb() })
        .pipe(gulp.dest(cssDestRoot))
}
```
if you print a debug message in each `on err` / `on end` place, you should find out that the `end` event always trigger before `error` event, that makes `error` just make no sense, because `cb` could only be call only once

if there's no error, both `end` callback will be called so finally `cb` called
 
if error occurred, only the first `end` callback was called, the second wouldn't, so `cb` never called, though `error` event could be emitted
